lib/src/iterable_extensions.dart

```diff
@@ -601,6 +601,19 @@ extension IterableExtension<T> on Iterable<T> {
       yield slice;
     }
   }
+
+  /// Returns a map where the keys are the unique elements of the iterable
+  /// and the values are the counts of those elements.
+  ///
+  /// For example, `['a', 'b', 'b', 'c', 'c', 'c'].countFrequency()`
+  /// returns `{'a': 1, 'b': 2, 'c': 3}`.
+  Map<T, int> countFrequency() {
+    var frequencyMap = <T, int>{};
+    for (var item in this) {
+      frequencyMap[item] = (frequencyMap[item] ?? 0) + 1;
```

Comment:
Can use `update`, which avoids doing two lookups per item.
```dart
   frequencyMap.update(item, _increment, ifAbsent: _one);
```
with top-level/static helpers:
```dart
  int _increment(int value) => value + 1;
  int _one() => 1;
```

Reply:
thank you, I agree this way is better, please check my changes

Reply:
How expensive is a lookup? Shouldn't it `O(1)`? So we are basically avoiding `O(2)`?

Reply:
Yes, we are avoiding O(2) and lowering the constant factor. I hope, that assumes the map is actually implemented efficiently, but it's at least *possible*.

Also, for degenerate cases, hash table lookup is linear. And quick-sort is quadratic. We always hope to not hit those cases, but when we do, it's better to have a low constant.

For library code that other people will use, especially low-level code like this, efficiency is important.
If all your helper libraries are 10% slower than necessary, then your entire application is *at least* 10% slower than necessary. (If helper libraries use helper libraries, it may compound.)
And there is nothing you can do about it.

So when you provide code for others to depend on, to build their algorithms and applications on, you should *never* say "a constant factor doesn't matter", because to end users, the difference between 15ms and 17ms may be a missed animation frame, and that *matters*.

(So, that's my pocket philosophy about writing code for others: The only person who can say whether performance is "good enough" is the application writer. A helper library doesn't know that. It may choose *how* it optimizes, for speed or size or a tradeoff, but it shouldn't say "it's good enough" and leave easy optimizations on the table. Every level below the application has to do their best and not be part of the problem.)

The one thing we should actually do is to *measure*. If `operator[]` and `operator[]=` are so heavily optimizied that they beat `update` even if they do the same lookup twice, then we need to optimize `update` too. I want to choose the operation that has the best *possible* performance, rather than using workarounds that are faster right now, and never come back and check if they stay that way. If `update` isn't faster today, we should file an issue, so that it is tomorrow.

(I can see that I flip between pragmatic and ideological reasoning - be fast, because it matters, but don't be fast by doing it the wrong way, even if it *is* faster today. Tradeoffs. It's tradeoffs all the way down. So reasonable people can disagree on where to trade what. )

Reply:
I'm not disagreeing, just wanted to understand the reasoning.

Thanks for the explanation.